src/core_ocean/shared/mpas_ocn_gm.F

```diff
@@ -196,323 +198,265 @@ subroutine ocn_gm_compute_Bolus_velocity(diagnosticsPool, meshPool, scratchPool)
+         dSdzTop(1) = -1e-15_RKIND
+         dzTop(maxLevelCell(iCell)+1) = -1e-15_RKIND
+         dTdzTop(maxLevelCell(iCell)+1) = -1e-15_RKIND
+         dSdzTop(maxLevelCell(iCell)+1) = -1e-15_RKIND
 
-      nCells = nCellsArray( 2 )
-
-      !$omp do schedule(runtime) private(i, iEdge, areaEdge, rtmp, k)
-      do iCell = 1, nCells
-         areaCellSum(:, iCell) = 1.0e-34_RKIND
          do i = 1, nEdgesOnCell(iCell)
             iEdge = edgesOnCell(i, iCell)
+            cell1 = cellsOnEdge(1,iEdge)
+            cell2 = cellsOnEdge(2,iEdge)
+            ! mrp Performance Note: Replace with a logical array on init later
+            if (cell1==iCell) then
+                iCellSelf = 1
+            elseif (cell2==iCell) then
+                iCellSelf = 2
+            else
+                print *, 'iCellSelf, something went wrong',cell1,cell2,iCell,nCells
+            endif
+            dcEdgeInv = 1.0_RKIND / dcEdge(iEdge)
+            areaEdge = dcEdge(iEdge) * dvEdge(iEdge)
+
+            if (maxLevelEdgeTop(iEdge) > 0) then  
+              indexMLDedge = max(1,indMLD(cell1), indMLD(cell2))
+              indexMLDedge = min(maxLevelEdgeTop(iEdge),indexMLDedge)
+            else
+              indexMLDedge = 1
+            endif
+            sshEdge = max(ssh(cell1), ssh(cell2))
+            zMLD = sshEdge - 0.5_RKIND*(zMid(indexMLDedge,cell1) + zMid(indexMLDedge,cell2))
 
-            !contribution of cell area from this edge * 2.0
-            areaEdge = 0.5_RKIND * dcEdge(iEdge) * dvEdge(iEdge)
             do k = 1, maxLevelEdgeTop(iEdge)
-               rtmp = areaEdge * relativeSlopeTopOfEdge(k, iEdge)**2
-               relativeSlopeTopOfCell(k, iCell) = relativeSlopeTopOfCell(k, iCell) + rtmp
-               areaCellSum(k, iCell) = areaCellSum(k, iCell) + areaEdge
-            end do
-         end do
-      end do
-      !$omp end do
+               drhoDT = -inSituThermalExpansionCoeff(k,iCell)
+               drhoDS = inSituSalineContractionCoeff(k,iCell)
+               dTdx = (  activeTracers(indexTemperature,k,cell2) &
+                       - activeTracers(indexTemperature,k,cell1) ) &
+                      * dcEdgeInv
+               dSdx = (  activeTracers(indexSalinity,k,cell2) &
+                       - activeTracers(indexSalinity,k,cell1) ) &
+                      * dcEdgeInv
+               drhoDx = drhoDT * dTdx + drhoDS * dSdx
+
+               ! Always compute *Up on the top cell and *Down on the bottom
+               ! cell, even though they are never used. This avoids an if
+               ! statement or separate computation for top and bottom.
+               slopeTriadUp(k,iCellSelf,iEdge) = &
+                  - drhoDx / &
+                  (  drhoDT * dTdzTop(k) &
+                   + drhoDS * dSdzTop(k) )
+               slopeTriadDown(k,iCellSelf,iEdge) = &
+                  - drhoDx / &
+                  (  drhoDT * dTdzTop(k+1) &
+                   + drhoDS * dSdzTop(k+1) )
```

Comment:
The QU240 gets a `floating invalid` on this line with a standard run (performance_test) after I turned on the MLD AM. I'm looking into it. @vanroekel can you run a standard test with debug on and see if it trips it? Here are my flags, in case there is some combination that might cause this:

<details>
<summary> flags </summary> <p>

```
in /lustre/scratch4/turquoise/mpeterse/runs/nightly/200402_rebase_rediFixesAndNewTapering_Luke_PR/ocean/global_ocean/QU240/performance_test/forward
 70 &Redi_isopycnal_mixing
 71     config_use_Redi = .true.
 72     config_Redi_kappa = 600.0
 73     config_Redi_maximum_slope = 0.3
 74     config_Redi_closure = 'constant'
 75     config_Redi_diabatic_limit = 5e-7
 76     config_Redi_use_slope_taper = .true.
 77     config_Redi_use_surface_taper = .true.
 78     config_Redi_N2_limit_term1 = .true.
 79 /
 80 &GM_eddy_parameterization
 81     config_use_GM = .true.
 82     config_GM_kappa = 600.0
 83     config_GM_closure = 'N2_dependent'
 84     config_GM_min_stratification_ratio = 0.1
 85     config_GM_constant_gravWaveSpeed = 0.3
 86 /
 87 &mesoscale_eddy_parameterizations
 88     config_eddying_resolution_taper = 'ramp'
 89     config_eddying_resolution_ramp_min = 20e3
 90     config_eddying_resolution_ramp_max = 30e3
```
</details>